src/ring_theory/tensor_product.lean

```diff
@@ -422,17 +422,19 @@ begin
   simp [H],
 end
 
--- TODO: with `smul_comm_class R S A` we can have this as an `S`-algebra morphism
-/-- The `R`-algebra morphism `A →ₐ[R] A ⊗[R] B` sending `a` to `a ⊗ₜ 1`. -/
-def include_left : A →ₐ[R] A ⊗[R] B :=
+/-- The `R`-algebra morphism `A →ₐ[S] A ⊗[R] B` sending `a` to `a ⊗ₜ 1`. -/
+def include_left [smul_comm_class R S A] : A →ₐ[S] A ⊗[R] B :=
```

Comment:
`S` should be explicit here

Reply:
You're right, and it's often painful otherwise (hence a few `include_left : _ \to\_a[S] _` later in the file). However, I wonder whether one could have a variant `include_left` where `S` and `R` are equal, and `include_left'` for the more general one (with one explicit argument).

Reply:
We do use that pattern elsewhere; though it means you end up duplicating lots of api

Reply:
I don't understand what you mean with “duplicating lots of api”: no new lemma is created.
In the commutative case, the application suggested by @ADedecker requires less than what is done here, namely just `A \to\_a A \otimes[R] B`. But if `A` is not commutative, that does not work and one needs all intermediate linearity properties `A \to\_a [S] A\otimes[R] B`, for adequate `S`.

Reply:
If we have both `include_left` and `include_left'`, then every time you write a lemma about the primed one you have to duplicate it for the unprimed one. So I think we should just have only the fully-general one; I don't think making people write a single `R` is much of a burden.

Reply:
Thank you, I see. I had imagined having `include_left := include_left' R`, but that would require many simp only [include_left] to access the API of `include_left'`, or duplicating it as you say.

Reply:
(I'll have to check, but it seems that one needs to do similar modifications in file#linear_map.tensor_product)

Reply:
I would suggest making this generalization in a separate PR; the typeclass change by itelf is pretty self-contained

Reply:
If I understand you correctly, this PR would just change the few instances that are invoked, and adding linearity is given in a separate PR ?